### PR TITLE
Text layout perf improvements

### PR DESCRIPTION
**Pull request by MrJul**
## What does the pull request do?
Improves the performance of text layout by using `ReadOnlySpan<T>` and `ReadOnlyMemory<T>` as much as possible. See https://github.com/AvaloniaUI/Avalonia/discussions/9978 for further discussion.

Notable changes:
 - `GlyphRun.Characters` is a `ReadOnlyMemory<char>` instead of a `IReadOnlyList<char>`.
 - `TextRun.Text` is a is a `ReadOnlyMemory<char>` instead of a `CharacterBufferReference`.
 - `CharacterBufferReference` and `CharacterBufferRange` are removed.

## Perf changes:

Current master:
|                Method |       Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Allocated |
|---------------------- |-----------:|---------:|---------:|---------:|--------:|----------:|
|       BuildTextLayout |   634.2 us |  5.67 us |  5.30 us |   0.9766 |       - |     25 KB |
| BuildEmojisTextLayout | 1,364.1 us | 12.15 us | 11.37 us |        - |       - |     30 KB |
|   BuildManySmallTexts | 4,043.9 us | 21.54 us | 20.15 us | 156.2500 | 78.1250 |  2,563 KB |
|  VirtualizeTextBlocks | 5,088.5 us | 20.43 us | 19.11 us | 148.4375 | 23.4375 |  2,445 KB |

This PR:
|                Method |       Mean |    Error |   StdDev |    Gen 0 |   Gen 1 | Allocated |
|---------------------- |-----------:|---------:|---------:|---------:|--------:|----------:|
|       BuildTextLayout |   355.3 us |  0.95 us |  0.89 us |   0.9766 |       - |     22 KB |
| BuildEmojisTextLayout |   735.4 us |  2.74 us |  2.56 us |   0.9766 |       - |     27 KB |
|   BuildManySmallTexts | 3,267.1 us | 10.25 us |  9.59 us | 144.5313 | 70.3125 |  2,391 KB |
|  VirtualizeTextBlocks | 4,291.4 us | 14.17 us | 12.56 us | 132.8125 | 15.6250 |  2,273 KB |

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0028645-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

File: tests/Avalonia.UnitTests/HarfBuzzTextShaperImpl.cs [hunk 1]
```diff
@@ -59,12 +62,21 @@ public ShapedBuffer ShapeText(CharacterBufferReference text, int textLength, Tex
 
                     var glyphIndex = (ushort)sourceInfo.Codepoint;
 
-                    var glyphCluster = (int)sourceInfo.Cluster;
+                    var glyphCluster = (int)(sourceInfo.Cluster);
 
-                    var glyphAdvance = GetGlyphAdvance(glyphPositions, i, textScale);
+                    var glyphAdvance = GetGlyphAdvance(glyphPositions, i, textScale) + options.LetterSpacing;
```
Comment (Mikolaytis) [inline tests/Avalonia.UnitTests/HarfBuzzTextShaperImpl.cs:67]: Nice fix

**Comment by Mikolaytis** [comment]
How much time do you invested into this?

**Comment by Mikolaytis** [comment]
carefully merged to Lunacy. Getting nice results:

before:
|                Method |       Mean |     Error |    StdDev |     Gen0 |     Gen1 |   Gen2 |  Allocated |
|---------------------- |-----------:|----------:|----------:|---------:|---------:|-------:|-----------:|
|       BuildTextLayout |   302.2 us |   6.58 us |  18.87 us |   3.9063 |   0.9766 |      - |   25.18 KB |
| BuildEmojisTextLayout |   273.4 us |   5.24 us |  10.70 us |   4.3945 |   1.9531 | 0.4883 |    28.7 KB |
|   BuildManySmallTexts | 5,353.1 us | 106.52 us | 263.28 us | 414.0625 | 203.1250 |      - | 2532.16 KB |

after:
|                Method |       Mean |    Error |    StdDev |     Gen0 |     Gen1 |   Gen2 | Allocated |
|---------------------- |-----------:|---------:|----------:|---------:|---------:|-------:|----------:|
|       BuildTextLayout |   238.2 us |  4.53 us |   4.23 us |   1.9531 |   0.9766 |      - |  12.15 KB |
| BuildEmojisTextLayout |   195.3 us |  3.66 us |   3.59 us |   1.7090 |   0.7324 | 0.2441 |  10.76 KB |
|   BuildManySmallTexts | 4,568.1 us | 90.70 us | 258.78 us | 390.6250 | 195.3125 |      - | 2368.1 KB |

Numbers are bit different from above because I've added some optimisations that not published yet.

**Comment by danipen** [comment]
I guess we can some perf improvements in AvaloniaEdit using these changes. Thanks @MrJul!

**Review by Gillibald — APPROVED** [review]
LGTM

Thanks a lot

**Pull request merged**